### PR TITLE
Make sure all pipelines can run with batched input

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR makes sure that all pipelines can be used with batched input. This usually means that all pipelines are tested both for single images and a list of pil images as well as a list of prompts of varying length.

Currently we have a bug in the main stable diffusion pipeline which should be solved by this PR: https://github.com/huggingface/diffusers/issues/1546

Slow tests were run for depth-to-image and image variation.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 1]
```diff
@@ -79,7 +90,7 @@ class StableDiffusionImg2ImgPipeline(DiffusionPipeline):
         feature_extractor ([`CLIPFeatureExtractor`]):
             Model that extracts features from generated images to be used as inputs for the `safety_checker`.
     """
-    _optional_components = ["safety_checker", "feature_extractor"]
+    _optional_components = ["safety_checker"]
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:93]: All pipelines that are conditioned have to use the feature extractor. We want in the short term to be able to move all the processing logic there instead of relying on handcrafted preprocessing functions like the one above.

File: src/diffusers/pipelines/stable_diffusion/pipeline_cycle_diffusion.py [hunk 2]
```diff
@@ -35,14 +35,26 @@
 logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
 
 
+# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.preprocess
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_cycle_diffusion.py:38]: Let's use already existing pre-processing functions

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -248,9 +248,9 @@ def _encode_prompt(self, prompt, device, num_images_per_prompt, do_classifier_fr
             return_tensors="pt",
         )
         text_input_ids = text_inputs.input_ids
-        untruncated_ids = self.tokenizer(prompt, padding="max_length", return_tensors="pt").input_ids
+        untruncated_ids = self.tokenizer(prompt, padding="longest", return_tensors="pt").input_ids
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:251]: This fixes: https://github.com/huggingface/diffusers/issues/1546

File: src/diffusers/pipelines/ddim/pipeline_ddim.py [hunk 4]
```diff
@@ -46,7 +46,6 @@ def __call__(
         use_clipped_model_output: Optional[bool] = None,
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
-        **kwargs,
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/ddim/pipeline_ddim.py:49]: never used

File: tests/test_pipelines_common.py [hunk 5]
```diff
@@ -94,7 +99,80 @@ def test_save_load_local(self):
...
+                continue
+            assert param in ALLOWED_REQUIRED_ARGS
+
+        optional_parameters = set({k for k, v in parameters.items() if v.default != inspect._empty})
+
+        required_optional_params = ["generator", "num_inference_steps", "return_dict"]
+        for param in required_optional_params:
+            assert param in optional_parameters
+
+    def test_inference_batch_consistent(self):
```
Comment (patrickvonplaten) [inline tests/test_pipelines_common.py:124]: That's a new test that makes sure all pipelines can be run in batch. In a future PR we should add another more aggressive test that checks that batched input always gives the same result is single input.

File: tests/test_pipelines_common.py [hunk 6]
```diff
@@ -94,7 +99,80 @@ def test_save_load_local(self):
...
+                        batched_inputs[name] = batch_size * [value]
+                elif name == "batch_size":
+                    batched_inputs[name] = batch_size
+                else:
+                    batched_inputs[name] = value
+
+            batched_inputs["num_inference_steps"] = inputs["num_inference_steps"]
+            batched_inputs["output_type"] = None
+
+            if self.pipeline_class.__name__ == "DanceDiffusionPipeline":
```
Comment (anton-l) [inline tests/test_pipelines_common.py:159]: Let's mark it in https://github.com/huggingface/diffusers/issues/1573 as well

**Review by anton-l — APPROVED** [review]
The tests and reusable processing functions are really nice, thanks for solving the tokenizer issue!

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 7]
```diff
@@ -323,8 +325,22 @@ def prepare_mask_latents(
...
 
         # duplicate mask and masked_image_latents for each generation per prompt, using mps friendly method
-        mask = mask.repeat(batch_size, 1, 1, 1)
-        masked_image_latents = masked_image_latents.repeat(batch_size, 1, 1, 1)
+        if mask.shape[0] < batch_size:
+            if not batch_size % mask.shape[0] == 0:
+                raise ValueError(
+                    "The passed mask and the required batch size don't match. Masks are supposed to be duplicated to"
+                    f" a total batch size of {batch_size}, but {mask.shape[0]} masks were passed. Make sure the number"
+                    " of images that you pass is divisble by the total requested batch size."
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:333]:
```suggestion
                    " of masks that you pass is divisible by the total requested batch size."
```

Comment (pcuenca) [reply src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:333]: Is this possible though, as the batch_size is inferred from the images passed?

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 8]
```diff
@@ -323,8 +325,22 @@ def prepare_mask_latents(
...
+                )
+            mask = mask.repeat(batch_size // mask.shape[0], 1, 1, 1)
+        if masked_image_latents.shape[0] < batch_size:
+            if not batch_size % masked_image_latents.shape[0] == 0:
+                raise ValueError(
+                    "The passed images and the required batch size don't match. Images are supposed to be duplicated"
+                    f" to a total batch size of {batch_size}, but {masked_image_latents.shape[0]} images were passed."
+                    " Make sure the number of images that you pass is divisble by the total requested batch size."
+                )
+            masked_image_latents = masked_image_latents.repeat(batch_size // masked_image_latents.shape[0], 1, 1, 1)
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:343]: This could be too much magic maybe? Why not accept either `1` mask, or the same as the number of images passed?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 9]
```diff
@@ -481,8 +496,22 @@ def prepare_mask_latents(
...
 
         # duplicate mask and masked_image_latents for each generation per prompt, using mps friendly method
-        mask = mask.repeat(batch_size, 1, 1, 1)
-        masked_image_latents = masked_image_latents.repeat(batch_size, 1, 1, 1)
+        if mask.shape[0] < batch_size:
+            if not batch_size % mask.shape[0] == 0:
+                raise ValueError(
+                    "The passed mask and the required batch size don't match. Masks are supposed to be duplicated to"
+                    f" a total batch size of {batch_size}, but {mask.shape[0]} masks were passed. Make sure the number"
+                    " of images that you pass is divisble by the total requested batch size."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:504]:
```suggestion
                    " of masks that you pass is divisible by the total requested batch size."
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 10]
```diff
@@ -481,8 +496,22 @@ def prepare_mask_latents(
...
+                )
+            mask = mask.repeat(batch_size // mask.shape[0], 1, 1, 1)
+        if masked_image_latents.shape[0] < batch_size:
+            if not batch_size % masked_image_latents.shape[0] == 0:
+                raise ValueError(
+                    "The passed images and the required batch size don't match. Images are supposed to be duplicated"
+                    f" to a total batch size of {batch_size}, but {masked_image_latents.shape[0]} images were passed."
+                    " Make sure the number of images that you pass is divisble by the total requested batch size."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:512]:
```suggestion
                    " Make sure the number of images that you pass is divisible by the total requested batch size."
```

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 11]
```diff
@@ -323,8 +325,22 @@ def prepare_mask_latents(
...
+                    f" a total batch size of {batch_size}, but {mask.shape[0]} masks were passed. Make sure the number"
+                    " of images that you pass is divisble by the total requested batch size."
+                )
+            mask = mask.repeat(batch_size // mask.shape[0], 1, 1, 1)
+        if masked_image_latents.shape[0] < batch_size:
+            if not batch_size % masked_image_latents.shape[0] == 0:
+                raise ValueError(
+                    "The passed images and the required batch size don't match. Images are supposed to be duplicated"
+                    f" to a total batch size of {batch_size}, but {masked_image_latents.shape[0]} images were passed."
+                    " Make sure the number of images that you pass is divisble by the total requested batch size."
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:341]:
```suggestion
                    " Make sure the number of images that you pass is divisible by the total requested batch size."
```

**Review by pcuenca — APPROVED** [review]

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:343]: Hmm, it's really related to this: https://github.com/huggingface/diffusers/pull/1669#discussion_r1047044968 -> happy to reiterate in a new PR. Think it's indeed an edge case that we sadly though already support

**Comment by patrickvonplaten** [comment]
Test failure seems flaky will investigate on "main"
